// @vkid/sdk/dist-sdk/cjs/index.js
'use strict';

var auth = require('./auth/auth.js');
var types$1 = require('./auth/types.js');
var config = require('./core/config/config.js');
var types = require('./core/config/types.js');
var widget = require('./core/widget/widget.js');
var events = require('./core/widget/events.js');
var types$2 = require('./types.js');
var oneTap = require('./widgets/oneTap/oneTap.js');
var types$3 = require('./widgets/oneTap/types.js');
var floatingOneTap = require('./widgets/floatingOneTap/floatingOneTap.js');
var types$4 = require('./widgets/floatingOneTap/types.js');
var oauthList = require('./widgets/oauthList/oauthList.js');
var types$5 = require('./widgets/oauthList/types.js');

var globalConfig = new config.Config();
/** Export Auth */ auth.Auth.config = globalConfig;
var globalAuth = new auth.Auth();
/** Export Core Widget */ widget.Widget.config = globalConfig;
widget.Widget.auth = globalAuth;

Object.defineProperty(exports, 'AuthErrorCode', {
  enumerable: true,
  get: function () { return types$1.AuthErrorCode; }
});
Object.defineProperty(exports, 'ConfigAuthMode', {
  enumerable: true,
  get: function () { return types.ConfigAuthMode; }
});
Object.defineProperty(exports, 'Prompt', {
  enumerable: true,
  get: function () { return types.Prompt; }
});
Object.defineProperty(exports, 'WidgetEvents', {
  enumerable: true,
  get: function () { return events.WidgetEvents; }
});
Object.defineProperty(exports, 'Languages', {
  enumerable: true,
  get: function () { return types$2.Languages; }
});
Object.defineProperty(exports, 'Scheme', {
  enumerable: true,
  get: function () { return types$2.Scheme; }
});
exports.OneTap = oneTap.OneTap;
Object.defineProperty(exports, 'OneTapSkin', {
  enumerable: true,
  get: function () { return types$3.OneTapSkin; }
});
exports.FloatingOneTap = floatingOneTap.FloatingOneTap;
Object.defineProperty(exports, 'FloatingOneTapContentId', {
  enumerable: true,
  get: function () { return types$4.FloatingOneTapContentId; }
});
exports.OAuthList = oauthList.OAuthList;
Object.defineProperty(exports, 'OAuthName', {
  enumerable: true,
  get: function () { return types$5.OAuthName; }
});
exports.Auth = globalAuth;
exports.Config = globalConfig;
